### PR TITLE
Define app-started payload object

### PR DESCRIPTION
**Pull request by jennchenn**
## Description
As the first step for the telemetry project, this PR creates a class that matches the `app-started` event payload, as well as helper classes that define integrations, dependencies and configurations/additional payloads. All schema details were taken from [here](https://github.com/DataDog/instrumentation-telemetry-api-docs/blob/main/GeneratedDocumentation/ApiDocs/v1/SchemaDocumentation/Schemas/app_started.md).

These classes will be used to format the data to be sent to the telemetry API. Parameter validation was added to ensure the class objects will be well-formed.

Some tests were added to confirm that the initialization of the objects works as expected, and that the validation checks run.

Questions:
* Does the naming make sense?

### CODE REVIEW

File: lib/datadog/core/telemetry/schemas/base/v1/application.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
+module Datadog
+  module Core
+    module Telemetry
+      module Schemas
+        module Base
+          module V1
+            # Describes attributes for application environment object
+            class Application
+              attr_reader :language_name, :language_version, :service_name, :tracer_version, :env, :runtime_name,
```
Comment (delner) [inline lib/datadog/core/telemetry/schemas/base/v1/application.rb:9]: Minor, but from a style point of view, when you have multiple attributes, it's easier to read them as a vertical list, e.g.

```ruby
attr_reader \
  :foo,
  :bar,
  :baz
```

File: lib/datadog/core/telemetry/schemas/base/v1/application.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+                          :runtime_patches, :runtime_version, :service_version, :products
+
+              def initialize(language_name, language_version, service_name, tracer_version, env = nil, runtime_name = nil,
```
Comment (delner) [inline lib/datadog/core/telemetry/schemas/base/v1/application.rb:12]: This argument list might be more appropriate as keyword arguments.

https://thoughtbot.com/blog/ruby-2-keyword-arguments

File: lib/datadog/core/telemetry/schemas/events/v1/app_started.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  module Core
+    module Telemetry
+      module Schemas
+        module Events
+          module V1
+            # Describes payload for telemetry V1 API app_started event
+            class AppStarted
+              attr_reader :configuration, :dependencies, :integrations, :additional_payload
+
+              def initialize(configuration = nil, dependencies = nil, integrations = nil, additional_payload = nil)
```
Comment (delner) [inline lib/datadog/core/telemetry/schemas/events/v1/app_started.rb:11]: I'm wondering if we should make some of these args required.

The general idea is we don't want poorly formed events; if you can build an event object, then its legit, so we don't have to validate it downstream. So these constructors should ensure the input is good.

This means we likely need validation logic implemented, as well as corresponding tests.

**Review by delner — COMMENTED** [review]
A good start! Object structure generally makes sense.

Left some feedback on some style points and other functional changes I think we'd like to see. Didn't give tests close scrutiny yet, but once the code changes are mostly complete, I'll give that a more thorough pass.

File: lib/datadog/core/telemetry/schemas/base/v1/application.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
+module Datadog
+  module Core
+    module Telemetry
+      module Schemas
+        module Base
+          module V1
+            # Describes attributes for application environment object
+            class Application
```
Comment (ivoanjo) [inline lib/datadog/core/telemetry/schemas/base/v1/application.rb:8]: Minor/more of a design thing: `Is::This::Class::Hierarchy::Getting::A::Bit::Too::Deep::And::Hard::To::Read`?

Such deep hierarchies are more common in Java, but they are seldom used in Ruby, because they can get really awkward to use (because of scoping rules, as well as not everyone using an IDE). 

For instance, even activesupport -- one of the most used libraries in Ruby and a core rails component ( https://github.com/rails/rails/tree/main/activesupport ) does not get 7 levels deep as we're doing here.

Could we perhaps shorten this to `Datadog::Core::Telemetry::Application`?

Comment (jennchenn) [reply lib/datadog/core/telemetry/schemas/base/v1/application.rb]: That makes sense - would keeping the version still be useful (i.e. something like `Datadog::Core::Telemetry::V1::Application`)?

Comment (ivoanjo) [reply lib/datadog/core/telemetry/schemas/base/v1/application.rb]: > That makes sense - would keeping the version still be useful (i.e. something like Datadog::Core::Telemetry::V1::Application)?

That is an excellent question! We can't tell the future and how things will evolve, so if we think it's likely that a V2 and possibly others would come along, keeping `V1` would make a lot of sense; if, on the other hand, that's not very likely, then we'll live with the `V1` in there for years without ever needing it.

My past experience tells me we are probably not going to need to use a V2. There is a version in the API itself, because APIs are often a bit more strictly specified and less flexible, but in code it's common that we can support all api versions with the same class. (And even on the backend side, a common technique is also to implement multiple versions of the api with the same code as well).

So... my suggestion would be to not have the `V1`, but since this is 🔮 futurology, it may turn out to be wrong. And that's fine, I think -- the worst case is that we'll need to introduce a `Telemetry::V2::Application` at some point and deprecate `Telemetry::Application`.

See also YAGNI -> https://martinfowler.com/bliki/Yagni.html :)

Comment (delner) [reply lib/datadog/core/telemetry/schemas/base/v1/application.rb]: I agree with the idea of flattening the namespacing a bit here: `Schemas` and `Base` are a bit superfluous. Categorically interesting, but not necessary. These can be removed.

I think we should keep `V1`: the spec for this already describes the telemetry schema as "V1", and there are plans to iterate. It especially makes sense when the API is in the agent, and we make need to upgrade/downgrade protocols. Though I can appreciate YAGNI, it's not applicable here, and we benefit from a little "futurology" when it prevents a breaking change.

**Review by marcotc — CHANGES_REQUESTED** [review]
A question to all: we are trying to type check Ruby across this PR.

Correct me if I'm wrong, but the classes introduced in this PR are used by Datadog only, not library users.
If so, I don't see the need for this explicit type validation on every `#initialize` call, as we control and will test all call sites.

All classes introduced in this PR look like [Structs](https://ruby-doc.org/core-2.5.0/Struct.html) with some light validation to me, but they read very verbosely.

~~IMO, we should remove any trivial validation and leave only product-related validations like around `API_VERSIONS` and `REQUEST_TYPES`.~~

IMO, we should remove all validation from the `#initialize` methods in this PR. I was on the fence about validating `API_VERSIONS` and `REQUEST_TYPES` but, `API_VERSIONS` is actually a constant (always `v1`) and `REQUEST_TYPES` is the responsibility of the caller (the caller will have a matching unit test to check if they set the expected type).

I think another way to summarize my sentiment is: I think the validation is being done in the wrong layer. I think it should be done by the caller that sets values into the data structures here.

File: lib/datadog/core/telemetry/schemas/base/v1/application.rb (hunk 4 above)
Comment (marcotc) [reply lib/datadog/core/telemetry/schemas/base/v1/application.rb]: Let me leave maybe a worse recommendation: `Datadog::Core::TelemetryV1`?

**Review by marcotc — APPROVED** [review]
